### PR TITLE
Add TextMarks to TextNode

### PR DESCRIPTION
**Pull request by trueadm**
This PR adds TextMarks to TextNodes. Specifically, it allows us to add marked ranges to specific text nodes, along with an `id` that can be used to reference external resources – such as comments, spell check data, etc. It could also be used to improve collaboration cursors.

There are 3 APIs introduced by this:

- `setMark(id, start, end)`
- `deleteMark(id)`
- `getMark(id)`

The API behaves like a `Map` but internally it is represented with an array. The reasoning for this implementation:

- Arrays can be serialized to JSON easily. If we used an exotic type like Map or Set, we'd have to add a `toJSON` to every type of `TextNode` which would be a massive breaking change.
- Objects are much slower to clone than arrays, Maps are faster than objects even. Furthermore, I don't ever anticipate there being that many marks on a single text node, and iterating through an array even up to a size of 32 elements is comparable to a Map lookup. If less than 8 elements, it's actually faster than a Map lookup in most cases.

This API design is to allow further exploration in the world of annotations/commenting/spell checking features for Lexical. Lastly, I went for "marks" because I see this as a way of marking a range, or part of a range, of the characters in a text node with an id that can be used externally. It felt a bit more descriptive than "data" or "metadata".

### CODE REVIEW

File: packages/lexical/src/nodes/LexicalTextNode.js [hunk 1]
```diff
@@ -253,13 +296,28 @@ export class TextNode extends LexicalNode {
...
   }
 
   getFormat(): number {
     const self = this.getLatest();
     return self.__format;
   }
 
+  getMark(id: string): null | TextMark {
+    const self = this.getLatest();
+    const marks = self.__marks;
```
Comment (fantactuka) [inline packages/lexical/src/nodes/LexicalTextNode.js:309]: Should it be `Map<id, Mark>` instead? Getting/setting will be faster, iteration about the same (unless there are plans to sort it be start/end)

**Comment by fantactuka** [comment]
Had another comment regarding map usage, thanks for details in the description 👍

File: packages/lexical/flow/Lexical.js.flow [hunk 2]
```diff
@@ -588,18 +588,25 @@ export type TextFormatType =
   | 'code'
   | 'subscript'
   | 'superscript';
-type TextModeType = 'normal' | 'token' | 'segmented' | 'inert';
+type TextMode = 'normal' | 'token' | 'segmented' | 'inert';
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/lexical/flow/Lexical.js.flow:591]: Did you check the overview? We can't really use a Map, and it doesn't really offer much anyway.

**Review by zurfyx — APPROVED** [review]
Sounds like the best approach from the options you've presented offline but it feels like a escape patch for our tree model, this resembles the old WWW approach where we had a long string with offsets to define various kinds of formatting

**Comment by zwily** [comment]
The limiting thing about this approach from an annotations/commenting perspective is that it limits the annotation/comment to refer to text inside a single text node, right? Or, you could add the same marks to adjacent nodes, but then you get into the problem of keeping those consistent in the face of node splits, etc.

(I think - this is all based on what little I understand of Lexical so far, and my background with Slate.)

**Comment by trueadm** [comment]
> The limiting thing about this approach from an annotations/commenting perspective is that it limits the annotation/comment to refer to text inside a single text node, right? Or, you could add the same marks to adjacent nodes, but then you get into the problem of keeping those consistent in the face of node splits, etc.

Internally, we adjust the marks when things are split etc automatically.

**Pull request merged**